src/com/facebook/buck/python/PythonRunTestsStep.java

```diff
@@ -0,0 +1,160 @@
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Optional;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.regex.Pattern;
+
+public class PythonRunTestsStep implements Step {
+
+  private final Path workingDirectory;
+  private final String testName;
+  private final ImmutableList<String> commandPrefix;
+  private final Supplier<ImmutableMap<String, String>> environment;
+  private final TestSelectorList testSelectorList;
+  private final Optional<Long> testRuleTimeoutMs;
+  private final Path resultsOutputPath;
+
+  private boolean timedOut;
+
+  public PythonRunTestsStep(
+      Path workingDirectory,
+      String testName,
+      ImmutableList<String> commandPrefix,
+      Supplier<ImmutableMap<String, String>> environment,
+      TestSelectorList testSelectorList,
+      Optional<Long> testRuleTimeoutMs,
+      Path resultsOutputPath) {
+    this.workingDirectory = workingDirectory;
+    this.testName = testName;
+    this.commandPrefix = commandPrefix;
+    this.environment = environment;
+    this.testSelectorList = testSelectorList;
+    this.testRuleTimeoutMs = testRuleTimeoutMs;
+    this.resultsOutputPath = resultsOutputPath;
+
+    this.timedOut = false;
+  }
+
+  @Override
+  public int execute(ExecutionContext context) throws IOException, InterruptedException {
+    int exitCode = doExecute(context);
+    if (timedOut) {
+      throw new HumanReadableException(
+          "Following test case timed out: " + testName + ", with exitCode: " + exitCode);
+    }
+    return exitCode;
+  }
+
+  private int doExecute(ExecutionContext context) throws IOException, InterruptedException {
+    if (testSelectorList.isEmpty()) {
+      return getShellStepWithArgs("-o", resultsOutputPath.toString()).execute(context);
+    }
+
+    ShellStep listStep = getShellStepWithArgs("-l");
+    int exitCode = listStep.execute(context);
+    if (timedOut || exitCode != 0) {
+      return exitCode;
+    }
+
+    ImmutableList.Builder<String> testsToRunRe = ImmutableList.builder();
+
+    String output = listStep.getStdout();
+    for (String strTestCase : output.split("\n")) {
+      String[] testCase = strTestCase.split("#", 2);
+      if (testCase.length != 2) {
+        throw new RuntimeException(String.format(
+            "Bad test case name from python runner: '%s'", strTestCase));
+      }
+
+      TestDescription testDescription = new TestDescription(testCase[0], testCase[1]);
+      if (testSelectorList.isIncluded(testDescription)) {
+        testsToRunRe.add(Pattern.quote(testDescription.toString()));
+      }
+    }
+
+    String testsToRunRegex = "^" + Joiner.on('|').join(testsToRunRe.build()) + "$";
+
+    return getShellStepWithArgs(
+        "-o", resultsOutputPath.toString(), "-r", testsToRunRegex).execute(context);
```

Comment:
True. Anyway, we can leave that for another time.

On Fri, Feb 5, 2016 at 2:20 PM Shawn Wilsher <notifications@github.com>
wrote:

> In src/com/facebook/buck/python/PythonRunTestsStep.java
> <https://github.com/facebook/buck/pull/646#discussion_r52080644>:
>
> > +      String[] testCase = strTestCase.split("#", 2);
> > +      if (testCase.length != 2) {
> > +        throw new RuntimeException(String.format(
> > +            "Bad test case name from python runner: '%s'", strTestCase));
> > +      }
> > +
> > +      TestDescription testDescription = new TestDescription(testCase[0], testCase[1]);
> > +      if (testSelectorList.isIncluded(testDescription)) {
> > +        testsToRunRe.add(Pattern.quote(testDescription.toString()));
> > +      }
> > +    }
> > +
> > +    String testsToRunRegex = "^" + Joiner.on('|').join(testsToRunRe.build()) + "$";
> > +
> > +    return getShellStepWithArgs(
> > +        "-o", resultsOutputPath.toString(), "-r", testsToRunRegex).execute(context);
>
> Note that argparse is python2.7+, and Buck supports python2.6+. If we did
> this, we'd have to import the package for 2.6.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/buck/pull/646/files#r52080644>.
>